Decorators WIP

pzuraq opened the pull request.

ljharb committed:
Editorial: add missing assertion to ApplyStringOrNumericBinaryOperator (#2282)

Fixes #2281

bakkot committed:
Editorial: split up IterationStatement (#2280)

bakkot committed:
Meta: bump ecmarkup to 6.0.0 (#2289)

jmdyck committed:
Editorial: Merge the RHSs of AsyncFunctionExpression (#2275)

The nonterminals:
- FunctionExpression
- ClassExpression
- GeneratorExpression
- AsyncGeneratorExpression

are all defined with a single RHS involving an optional BindingIdentifier.

But AsyncFunctionExpression is defined with two RHSs, one with a BindingIdentifier and one without.
(It's been that way since it was introduced in PR #692)
I can't see any reason for it to be not like the others.

mathiasbynens committed:
Editorial: improve “About this Specification” section (#2284)

Published snapshots such as https://262.ecma-international.org/11.0/ include the exact same text. This patch tweaks the text so that the sentence still makes sense even when it’s being read from a snapshot copy. It also hardcodes the link’s target URL so that the link text and destination are always in sync and point to the expected location.

DanielRosenwasser committed:
Editorial: swap phrasing to indefinite article for Boolean-coercibles (#2295)

ljharb committed:
Editorial: add <dfn> for "TypedArray element type" (#2297)

Commit:
Editorial: convert Number abstract operations to algorithmic steps (#2189)

Co-authored-by: Ryan Duffy <ryan.duffy@lge.com>
Co-authored-by: Michael Ficarra <mficarra@shapesecurity.com>
Co-authored-by: Kevin Gibbons <bakkot@gmail.com>

bakkot committed:
Meta: bump ecmarkup to 6.0.1 (#2301)

bakkot committed:
Editorial: replace a "<=" with a "≤" (#2301)

szuend committed:
Normative: A more precise Array.prototype.sort (#1585)

Co-authored-by: Simon Zünd <szuend@chromium.org>
Co-authored-by: Mathias Bynens <mathias@qiwi.be>

rkirsling committed:
Normative: Give %TypedArray% methods explicit algorithms (#2221)

ljharb committed:
Editorial: remove unnecessary assertion (#2296)

Both S and R are Strings, so either both or neither should be asserted in the steps. I chose neither.

ExE-Boss committed:
Normative: Set function `name` and `length` in `CreateBuiltinFunction` (#2116)

Co-authored-by: Kevin Gibbons <bakkot@gmail.com>
Co-authored-by: Shu-yu Guo <syg@chromium.org>

ExE-Boss committed:
Editorial: Improve callback variable names used by `Promise` statics (#2116)

jmdyck committed:
Editorial: Standardize preambles for host-defined abstract operations (#2253)

... using the format established in PR #1914.

ljharb committed:
Markup: add missing `</p>` tag from 346e583 / #2116

bakkot committed:
Meta: bump ecmarkup to 6.1.0 (#2311)

bakkot committed:
Editorial: dfn Normative Optional (#2311)

syg committed:
Editorial: Remove paragraph listing similarities to other languages (#2312)

ljharb committed:
Meta: add `scroll-margin-top` CSS to make for a nicer anchor experience (#2317)

See https://piccalil.li/quick-tip/add-scroll-margin-to-all-elements-which-can-be-targeted

ExE-Boss committed:
Editorial: Fix `<dfn>`s for `WeakRef` and `FinalizationRegistry` (#2318)

bakkot committed:
Normative: forbid for-of loops with variable named async (#2256)

ljharb committed:
Editorial: swap "about" and "contributing" sections in the intro (#2324)

ljharb committed:
Editorial: nest ASI sections under "ASI" section (#2323)

devsnek committed:
Normative: Define default constructors using spec steps (#2216)

shvaikalesh committed:
Editorial: Align checks order in [[DefineOwnProperty]] overrides with ordinary method (#2333)

Commit:
Editorial: Fix mistakes (#2331)

jmdyck committed:
Editorial: Change "that satisfies" to "such that" (#2330)

The spec has only one occurrence of:
    "For each ... that satisfies ..."
but six of:
    "For each ... such that ..."
so change the former to the latter.

jmdyck committed:
Editorial: Change "non-negative integer" to "integer" (#2330)

... in the only occurrence of "For each non-negative integer",
to match the 4 existing occurrences of "For each integer".

jmdyck committed:
Editorial: Drop "in original [key] insertion order" (#2330)

The form
    "For each [element] of [list], in original key insertion order"
(or "... original insertion order") is weird because it implies that
it's not enough to know the order of the elements in the List,
you also have to know the order that they were inserted into the List.
But in fact, that's not necessary: in every case where this form is used,
the insertion order is the *same* as the List order,
because insertions only happen via appending to the List.

Map.prototype.forEach:
  Iterates over _map_.[[MapData]],
  whose only inserts are appends, in Map.prototype.set.

Set.prototype.forEach:
  Iterates over _set_.[[SetData]],
  whose only inserts are appends, in Set.prototype.add.

TriggerPromiseReactions:
  Iterates over _reactions_, which is
  _promise_.[[PromiseFulfillReactions]] or
  _promise_.[[PromiseRejectReactions]],
  whose only inserts are appends, in PerformPromiseThen.

bakkot committed:
Meta: better compression during preview uploads (#2340)

ljharb committed:
Editorial: add ES2021 intro text

ljharb committed:
Editorial: master is now ES2022

ljharb committed:
Meta: update editors list for ES2022

mathiasbynens committed:
Editorial: Clarify ES2021 Array.prototype.sort change (#2342)

`Array.prototype.sort` was already made stable in ECMAScript 2019 (see two paragraphs above). The change in https://github.com/tc39/ecma262/pull/1585 made `Array.prototype.sort` more precise, reducing the amount of cases that result in an implementation-defined sort order.

bakkot committed:
Normative: specify creation order for capturing group properties (#2329)

bakkot committed:
Meta: bump ecmarkup to 7.0.2 (#2339)

bakkot committed:
Meta: multipage builds (#2339)

syg committed:
Normative: Use HostJobCallbacks for FinalizationRegistry callbacks (#2316)

This aligns FinalizationRegistry callbacks with Promise callbacks in
allowing web browsers to perform additional steps to save and restore
the correct incumbent settings object.

jmdyck committed:
Editorial: Have NewGlobalEnvironment invoke NewObjectEnvironment (#2287)

... rather than creating "a new object Environment Record" directly.

(Now, all object Environment Records are created via NewObjectEnvironment.)

jmdyck committed:
Editorial: Set _withEnvironment_ explicitly (#2287)

... when creating an object Environment Record.

(So _withEnvironment_ no longer needs to have a "default" value.)

jmdyck committed:
Editorial: Add parameter to NewObjectEnvironemt (#2287)

... that gives the value for the new environment's _withEnvironment_ flag.

jmdyck committed:
Editorial: Introduce fields to specify object Env Rec (#2287)

Specifically: [[BindingObject]] and [[IsWithEnvironment]]

jmdyck committed:
Editorial: Rename alias `_bindings_` to `_bindingObject_` (#2287)

michaelficarra committed:
Editorial: use target alias name instead of "it" (#2341)

bakkot committed:
Meta: bump ecmarkup to 7.0.3 (#2344)

mathiasbynens committed:
Editorial: remove redundant `index.html` (#2344)

This is a follow-up to #2339.

michaelficarra committed:
Editorial: clean up wording of Array and TypedArray methods (#2346)

jmdyck committed:
Editorial: 5 simple tweaks that I believe are self-explanatory (#2351)

jmdyck committed:
Editorial: Change "that is the same sign as _number_" (#2351)

to "whose sign is the sign of _number_"

(My first thought was to simply change "is" to "has":
`the mathematical value that *has* the same sign as _number_ ...`
but then "and whose magnitude ..." still sounded odd,
so instead I'm suggesting a "whose sign" clause to balance it.)

jmdyck committed:
Editorial: Drop Script Record's [[Environment]] field (#2351)

It's set to `*undefined*` in ParseScript and then never referenced again.
(Instead, we have _scriptRecord_.[[Realm]].[[GlobalEnv]])

jmdyck committed:
Editorial: Change "headings" to "heading" (#2351)

I believe the spec referred to "headings" in
   "subclause headings for the function description"
to accommodate a few functions that were defined
via 'overloads' in multiple clauses.

But PR #2176 rewrote those functions so that each has
a single definition, and thus only one relevant header.

jmdyck committed:
Editorial: Change "includes" to "contains" (#2351)

I think this is the only (remaining) use of "include" with this sense.

See https://github.com/tc39/ecma262/pull/1554#discussion_r577744595

h2oche committed:
Editorial: Fix early error rules of 'IfStatement' (#2359)

jmdyck committed:
Editorial: Unfold SDO `CoveredParenthesizedExpression` (#2350)

jmdyck committed:
Editorial: Unfold SDO `CoveredCallExpression` (#2350)

jmdyck committed:
Editorial: Unfold SDO `CoveredAsyncArrowHead` (#2350)

jmdyck committed:
Editorial: Eliminate one use of SDO `CoveredFormalsList` (#2350)

Here in InstantiateArrowFunctionExpression,
I assert that we can replace the status quo's
    CoveredFormalsList of |ArrowParameters|
with just
    |ArrowParameters|
(and then eliminate the _parameters_ alias).

case analysis on the 2 forms of |ArrowParameters|:

case 1:
When |ArrowParameters| is an instance of
    ArrowParameters : BindingIdentifier
then CoveredFormalsList simply returns |ArrowParameters| itself,
so the above replacement is obviously correct.

case 2:
When |ArrowParameters| is an instance of
    ArrowParameters : CoverParenthesizedExpressionAndArrowParameterList
then CoveredFormalsList will return an instance of
    ArrowFormalParameters
(whose particulars are not important here),
which then gets passed to OrdinaryFunctionCreate's ParameterList.

But if you go to OrdinaryFunctionCreate and follow what happens
to ParameterList, you find that, ultimately, it's submitted to:
    - ExpectedArgumentCount
    - BoundNames
    - IsSimpleParameterList
    - ContainsExpression
    - IteratorBindingInitialization

And all of these have an explicit definition for:
    ArrowParameters : CoverParenthesizedExpressionAndArrowParameterList
of roughly the form:
    1. Let _formals_ be the CoveredFormalsList of |CoverParenthesizedExpressionAndArrowParameterList|.
    1. Return <SDO> of _formals_.

That is, with respect to the OrdinaryFunctionCreate's call-tree,
if you don't call CoveredFormalsList near the root,
it will be called near the leaves.

So in this case too, it's okay to perform the given replacement.

jmdyck committed:
Editorial: Unfold SDO `CoveredFormalsList` (#2350)

h2oche committed:
Editorial: Consistent notation of newly created error object (#2363)

jmdyck committed:
Editorial: Factor out abstract operation RegExpHasFlag (#2349)

h2oche committed:
Editorial: Correct usage of 'Let'/'Set' in various algorithms (#2365)

jmdyck committed:
Editorial: Simplify CreateDynamicFunction (#2348)

Formerly, CreateDynamicFunction would parse the text for the parameters and body *separately*, so we didn't have a complete function decl/expr, so the Early Error rules for the function decl/expr production didn't apply automatically, so we had to apply them "manually".

Instead, parse the text of a complete function expr (which text already exists in _sourceText_!), so that all the relevant Early Error rules apply automatically.

h2oche committed:
Editorial: Fix Contains algorithm of ClassTail (#2362)

Commit:
Add Class and Class Element Decorators and `accessor` Keyword

bakkot committed:
Editorial: suppress spurious linking of "Contains" (#2369)

bakkot committed:
Editorial: more explicit validation in CreateDynamicFunction (#2374)

Ms2ger committed:
Normative: Add public and private class fields, static class features, private methods & getters/setters

This includes changes from https://tc39.es/proposal-class-fields/.
This includes changes from https://tc39.es/proposal-private-methods/.
This includes changes from https://tc39.es/proposal-static-class-features/.

Co-authored-by: Ms2ger <Ms2ger@gmail.com>
Co-authored-by: Ujjwal Sharma <ryzokuken@igalia.com>
Co-authored-by: Caio Lima <ticaiolima@gmail.com>
Co-authored-by: Michael Dyck <jmdyck@ibiblio.org>
Co-authored-by: Kevin Gibbons <bakkot@gmail.com>
Co-authored-by: Jordan Harband <ljharb@gmail.com>

ljharb commented on spec.html:
```diff
@@ -3537,6 +3548,13 @@ <h1>The List and Record Specification Types</h1>
       <p>Schema for commonly used Record field combinations may be named, and that name may be used as a prefix to a literal Record value to identify the specific kind of aggregations that is being described. For example: PropertyDescriptor { [[Value]]: 42, [[Writable]]: *false*, [[Configurable]]: *true* }.</p>
     </emu-clause>
 
+    <emu-clause id="sec-map-specification-type">
+      <h1>The Map Specification Types</h1>
+      <p>The <dfn>Map</dfn> type is used to explain the evaluation of key-value based data structures, such as the metadata data structure built up by decorators (see <emu-xref href="#sec-metadata-record-specification-type"></emu-xref>). Values of the Map type are ordered sequences of key/value pairs containing individual values. These sequences may be of any length. The elements of a map may be  ...
+
+      <p>Maps can be iterated like lists, and are iterated as key/value pairs. If no order is specified, then elements are iterated in the order they were inserted.</p>
+    </emu-clause>
```
why do we need this type at all? couldn't we make a null object (one that's never observed to exist) and use that for the lookups?

ljharb commented on spec.html:
```diff
@@ -4206,6 +4224,28 @@ <h1>The ClassFieldDefinition Record Specification Type</h1>
+              [[DecoratorInitializers]]
+            </td>
+            <td>
+              a list of ECMAScript function objects, or ~empty~
```
```suggestion
              a List of ECMAScript function objects, or ~empty~
```
any reason not to default it to an empty List?

ljharb commented on spec.html:
```diff
@@ -4206,6 +4224,28 @@ <h1>The ClassFieldDefinition Record Specification Type</h1>
+              [[FinisherInitializers]]
+            </td>
+            <td>
+              a list of ECMAScript function objects, or ~empty~
```
```suggestion
              a List of ECMAScript function objects, or ~empty~
```
same question

ljharb commented on spec.html:
```diff
@@ -4222,6 +4262,85 @@ <h1>The ClassFieldDefinition Record Specification Type</h1>
+              [[Initializers]]
+            </td>
+            <td>
+              a list of ECMAScript function objects, or ~empty~
```
```suggestion
              a List of ECMAScript function objects, or ~empty~
```
here also

jmdyck committed:
Preserve defunct ids via 'oldids'

jmdyck committed:
Change literal angle quotes to HTML entities.

See PR #2149.

jmdyck committed:
Third argument to Call() is a List

jmdyck committed:
emu-clause tags: shorten ids, add type="sdo"

jmdyck committed:
abstract op preambles: standard wording

jmdyck committed:
Capitalization

jmdyck committed:
? and !

jmdyck committed:
algorithm syntax

jmdyck committed:
'type' of _metadataList_

jmdyck committed:
References to undefined aliases.

Please check that I've made the correct replacement!

jmdyck committed:
References to undefined operations

Please check that I've made the correct replacements

jmdyck committed:
Presumably PrivateValueSetter will be called with a value

jmdyck committed:
I'm pretty sure this '_metadataObj_' should be '_metadataObjForKey_'

(otherwise _metadataObjForKey_ wouldn't always be defined subequently)

jmdyck committed:
Typos

jmdyck committed:
PrivateSet doesn't return a value

or at least, not one that a function should be returning.

jmdyck committed:
Ensure that a SetDecoratorMetadata function returns a value

jmdyck committed:
Move up the Let for _staticMetadataList_

so that it's defined for the call to ClassElementEvaluation

jmdyck committed:
Add _initializers_ parameter to CreateDecoratorContextObject

Every call to CreateDecoratorContextObject assumes that this parameter is there.

(Of course, you'll need to add steps to *do* something with it.)

jmdyck committed:
Swap args in call to CreateDecoratorContextObject

jmdyck committed:
Change CreateDecoratorAccessObject arg

... to conform to what it accepts.

jmdyck committed:
Reference new productions in Annex A

pzuraq committed:
Merge pull request #2 from jmdyck/2417_ed

suggested changes

pzuraq commented:
Moved to https://github.com/tc39/ecma262/pull/2417
